### DISCUSSION TITLE
Power Flow algorithm

### DISCUSSION THREAD

**Discussion by YaeSakura1** [CURENT/andes#172] [Q&A] [answered]
Dr. Cui,
To solve the AC/DC power flow need sequential or unifified methods , how does ANDES do AC/DC power flow calculations? 
Thanks!

Best regards

**Comment by cuihantao** [comment, marked as answer]
Hi,

ANDES uses the unified. AC and DC devices are solved simultaneously based on the power injection formulation.

**Reply by YaeSakura1** [reply]
Thank you Dr.Cui,
Thanks for your reply!

Best regards

**Comment by YaeSakura1** [comment]
Dr. Cui,
To solve the LCC-VSC AC/DC power flow, i modeled the LCC model.
I want to know is that right? please give me some advice.
Thanks!

Best regards


"""line commutated converter models"""

from andes.models.dcbase import ACDC2Term
from andes.core.param import NumParam
from andes.core.var import Algeb, State, ExtState, ExtAlgeb  # NOQA
from andes.core.service import ConstService, ExtService  # NOQA
from andes.core.discrete import HardLimiter, Switcher, AntiWindup  # NOQA
from andes.core.block import PIController  #NOQA

class LCC(ACDC2Term):
    """Data for LCC in power flow"""
    def __init__(self, system, config):
        ACDC2Term.__init__(self, system, config)
        self.rsh = NumParam(default=0.0025, info="AC interface resistance", unit="ohm", z=True,
                            tex_name='r_{sh}')
        self.xsh = NumParam(default=0.06, info="AC interface reactance", unit="ohm", z=True,
                            tex_name='x_{sh}')

        self.p0 = NumParam(default=0.0, info="AC active power setting", unit="pu")
        self.q0 = NumParam(default=0.0, info="AC reactive power setting", unit="pu")
        self.vdc0 = NumParam(default=1.0, info="DC voltage setting", unit="pu", tex_name='v_{dc0}')

        self.k0 = NumParam(default=0.0, info="Loss coefficient - constant")
        self.k1 = NumParam(default=0.0, info="Loss coefficient - linear")
        self.k2 = NumParam(default=0.0, info="Loss coefficient - quadratic")

        self.droop = NumParam(default=0.0, info="Enable dc voltage droop control", unit="boolean")
        self.K = NumParam(default=0.0, info="Droop coefficient")
        self.vhigh = NumParam(default=9999, info="Upper voltage threshold in droop control", unit="pu")
        self.vlow = NumParam(default=0.0, info="Lower voltage threshold in droop control", unit="pu")

        self.vshmax = NumParam(default=1.1, info="Maximum ac interface voltage", unit="pu")
        self.vshmin = NumParam(default=0.9, info="Minimum ac interface voltage", unit="pu")
        self.Ishmax = NumParam(default=2, info="Maximum ac current", unit="pu")

        self.KR = NumParam(default=1, info="Rectifier-side transformer tap ratio", unit="pu/pu")
        self.KI = NumParam(default=1, info="Inverter-side transformer tap ratio", unit="pu/pu")

        self.amax = NumParam(default=100, info="Maximum firing angle", unit="rad")
        self.amin = NumParam(default=5, info="Minimum firing angle", unit="rad")
        self.bmax = NumParam(default=60, info="Maximum extinction angle", unit="rad")
        self.bmin = NumParam(default=15, info="Minimum extinction angle", unit="rad")

        self.Xc = NumParam(default=0.03, info="Commutation reactance", unit="pu")
        self.kr = NumParam(default=0.995, info="consider the commutation effect")
        self.N = NumParam(default=6, info="the converter bridge number")
        self.d = NumParam(default=0.436, info="the converter power factor angle", unit="rad")

        self.ids = NumParam(default=2, info="the current set value", unit="pu")
        self.vds = NumParam(default=1, info="the voltage set value", unit="pu")
        self.pds = NumParam(default=2, info="the power set value", unit="pu")
        self.ads = NumParam(default=18, info="the firing angles set value", unit="rad")
        self.bds = NumParam(default=22, info="the extinct angles set value", unit="rad")
        self.kds = NumParam(default=1, info="the ratios set value", unit="pu")

        self.ar = NumParam(default=15, info="firing delay angle", unit="rad")
        self.bi = NumParam(default=15, info="extinction advance angle", unit="rad")

        self.idr = NumParam(default=1)
        self.idi = NumParam(default=1)
        self.R = NumParam(default=3.6)

        self.control = NumParam(info="Control method: 0-CC, 1-CV, 2-CP, 3-CIA, 4-CEA, or 5-CR", mandatory=True)

        # define variables and equations
        self.flags.update({'pflow': True})
        self.flags.update({'tds': True})
        self.group = 'StaticLCC'

        self.mode = Switcher(u=self.control, options=(0, 1, 2, 3, 4, 5))


# 整流器方程
        self.vd0r = Algeb(info='ideal no-load direct voltage',
                          unit='p.u.',
                          tex_name='V_{d0r}',
                          v_str='vd0r',
                          e_str='3 * 2**0.5/3.14 * KR * N * v - vd0r',
                          diag_eps=True,
                          )
        self.vdr = Algeb(info='rectifier dc voltage',
                         unit='p.u.',
                         tex_name='V_{dr}',
                         v_str='vdr',
                         e_str='3 * 2**0.5/3.14 * KR * N * v * cos(ar)- 3/3.14 * N * Xc * idr - vdr',
                         diag_eps=True,
                         )
        self.pr = Algeb(info='rectifier active power',
                        unit='p.u.',
                        tex_name='P_{r}',
                        v_str='pr',
                        e_str='vdr * idr - pr',
                        diag_eps=True,
                        )
        self.qr = Algeb(info='rectifier reactive power',
                        unit='p.u.',
                        tex_name='Q_{r}',
                        v_str='qr',
                        e_str='(vdr * idr) * tan(dr) - qr',
                        diag_eps=True,
                        )
        self.dr = Algeb(info='the phase angle between the ac voltage and fundamental current',
                        unit='p.u.',
                        tex_name=r'\phi_{r}',
                        v_str='dr',
                        e_str='acos(vdr/vd0r) - dr',
                        diag_eps=True,
                        )
# 逆变器方程
        self.vd0i = Algeb(info='ideal no-load direct voltage',
                          unit='p.u.',
                          tex_name='V_{d0i}',
                          v_str='vd0i',
                          e_str='3 * 2**0.5/3.14 * KR * N * v - vd0i',
                          diag_eps=True,
                          )
        self.vdi = Algeb(info='inverter dc voltage',
                         unit='p.u.',
                         tex_name='V_{di}',
                         v_str='vdi',
                         e_str='3 * 2**0.5/3.14 * KR * N * v * cos(bi)- 3/3.14 * N * Xc * idi - vdi',
                         diag_eps=True,
                         )
        self.pi = Algeb(info='inverter active power',
                        unit='p.u.',
                        tex_name='P_{i}',
                        v_str='pi',
                        e_str='vdi * idi - pi',
                        diag_eps=True,
                        )
        self.qi = Algeb(info='inverter reactive power',
                        unit='p.u.',
                        tex_name='Q_{i}',
                        v_str='qi',
                        e_str='(vdi * idi) * tan(di) - qi',
                        diag_eps=True,
                        )
        self.di = Algeb(info='the phase angle between the ac voltage and fundamental current',
                        unit='p.u.',
                        tex_name=r'\phi_{i}',
                        v_str='di',
                        e_str='acos(vdi/vd0i) - di',
                        diag_eps=True,
                        )

# 线路方程
        self.id = Algeb(info='dc current',
                        unit='p.u.',
                        tex_name='I_{d}',
                        v_str='id',
                        e_str='vdr - vdi - R * id',
                        diag_eps=True,
                        )
# 控制方程:
        self.CC = Algeb(info='constant current control',
                        unit='p.u.',
                        v_str='CC',
                        e_str='id - ids',
                        diag_eps=True,
                        )
        self.CV = Algeb(info='constant voltage control',
                        unit='p.u.',
                        v_str='CV',
                        e_str='id - vds',
                        diag_eps=True,
                        )
        self.CP = Algeb(info='constant power control',
                        unit='p.u.',
                        v_str='CP',
                        e_str='id * id - pds',
                        diag_eps=True,
                        )
        self.CIA = Algeb(info='constant firing angle control',
                         unit='raf',
                         v_str='CIA',
                         e_str='cos(ar) - cos(ads)',
                         diag_eps=True,
                         )
        self.CEA = Algeb(info='constant extinction angle control',
                         unit='rad',
                         v_str='CEA',
                         e_str='cos(bi) - cos(bds)',
                         diag_eps=True,
                         )
        self.CR = Algeb(info='constant ratios control',
                        unit='p.u.',
                        v_str='CR',
                        e_str='KR - kds',
                        diag_eps=True,
                        )

**Comment by YaeSakura1** [comment]
After i run a case, i get the following error:

C:\Users\Administrator\.andes\pycode\LCC.py:13: RuntimeWarning: invalid value encountered in true_divide
  return ((1.35115945449659*KR*N*v, 1.35115945449659*KR*N*v*cos(ar) - 0.955414012738854*N*Xc*idr, idr*vdr, idr*vdr*tan(dr), -dr + arccos(vdr/vd0r), 1.35115945449659*KR*N*v, 1.35115945449659*KR*N*v*cos(bi) - 0.955414012738854*N*Xc*idi, idi*vdi, idi*vdi*tan(di), arccos(vdi/vd0i), -R*id - vdi + vdr, id - ids, id - vds, id**2 - pds, -cos(ads) + cos(ar), -cos(bds) + cos(bi), KR - kds, 0, 0, 0, 0))
C:\Users\Administrator\.andes\pycode\LCC.py:16: RuntimeWarning: invalid value encountered in true_divide
  return ((1.35115945449659*KR*N, 1.35115945449659*KR*N*cos(ar), idr, idr*tan(dr), idr*vdr*(tan(dr)**2 + 1), vdr/(vd0r**2*sqrt(1 - vdr**2/vd0r**2)), -1/(vd0r*sqrt(1 - vdr**2/vd0r**2)), -1, 1.35115945449659*KR*N, 1.35115945449659*KR*N*cos(bi), idi, idi*tan(di), idi*vdi*(tan(di)**2 + 1), vdi/(vd0i**2*sqrt(1 - vdi**2/vd0i**2)), -1/(vd0i*sqrt(1 - vdi**2/vd0i**2)), 1, -1, -R, 1, 1, 2*id))
-> Single process finished in 1.0984 seconds.

**Comment by cuihantao** [comment]
You might have errors in the equations. One issue with your implementation is that you will need to separate the power flow model from the time-domain model. That is, use the power flow results to initialize the dynamic model. Not sure if it is implemented this way as I only see one class.

The error you are seeing is probably a division by zero. You can use a debugger to place breakpoints before entering the specific line in `LCC.py` and inspect the values.

**Comment by YaeSakura1** [comment]
Dr. Cui,

self.ash = Algeb(info='voltage phase behind the transformer',
                         unit='rad',
                         tex_name=r'\theta_{sh}',
                         v_str='a',
                         e_str='u * (gsh * v**2 - gsh * v * vsh * cos(a - ash) - '
                               'bsh * v * vsh * sin(a - ash)) - psh',
                         diag_eps=True,
                         )
What does v_str='a' and e_str= 'u * (gsh * v**2 - gsh * v * vsh * cos(a - ash) - '
                               'bsh * v * vsh * sin(a - ash)) - psh'  mean?
Is the v_str means the value string?
Is the e_str means : −𝑃𝑠ℎ+𝑢(𝑉𝑠ℎ𝑏𝑠ℎ𝑣sin(𝜃𝑠ℎ−𝑎)−𝑉𝑠ℎ𝑔𝑠ℎ𝑣cos(𝜃𝑠ℎ−𝑎)+𝑔𝑠ℎ𝑣2) = 0 ?

Thanks!
Best regards

**Comment by jinningwang** [comment]
Hi YaeSakura1,

This is Jinning.

1) You can use Markdown code format ("```") to display code block better.

2) `v_str` is the variable initialized values, and `e_str` is the variable equations on the RHS.
For example, if
```
self.A = Algeb(v_str='A0',
               e_str='B - A')
```
It will result in:
B - A = 0
where A = A0 at initial.

**Reply by YaeSakura1** [reply]
Thank you Jinning,
Thanks for your reply!

Best regards

**Comment by YaeSakura1** [comment]
`self.psh = Algeb(info='active power injection into VSC',
tex_name="P_{sh}",
unit='p.u.',
v_str='p0 * (mode_s0 + mode_s1)',
e_str='u * (mode_s0 + mode_s1) * (p0 - psh) + '
'u * (mode_s2 + mode_s3) * (v1 - v2 - vdc0)',
diag_eps=True,
)`
`self.qsh = Algeb(info='reactive power injection into VSC',
tex_name="Q_{sh}",
v_str='q0 * (mode_s0 + mode_s2)',
e_str='u * (mode_s0 + mode_s2) * (q0 - qsh) + '
'u * (mode_s1 + mode_s3) * (v0 - v)',
diag_eps=True,
)`

what does psh `e_str` and qsh `e_str` mean?

I print the mode_s0, mode_s1, mode_s2, mode_s3, and get
[1. 0.], [0. 0.], [0. 1.], [0. 0.]
how it work in the equation? and how to represent different controls?

**Comment by cuihantao** [comment]
The mode indicators are used in the equation to construct piece-wise equations.

**Comment by YaeSakura1** [comment]
I notice that you consider the loss coefficient in
`self.k0 = NumParam(default=0.0, info="Loss coefficient - constant")
self.k1 = NumParam(default=0.0, info="Loss coefficient - linear")
self.k2 = NumParam(default=0.0, info="Loss coefficient - quadratic")`

but seems that it doesn't show up in the equation, how does it work?

**Reply by cuihantao** [reply]
Looks like it has not been implemented yet..

**Reply by YaeSakura1** [reply]
so this VSC model is considered lossless?

`self.rsh = NumParam(default=0.0025, info="AC interface resistance", unit="ohm", z=True,tex_name='r_{sh}')
self.xsh = NumParam(default=0.06, info="AC interface reactance", unit="ohm", z=True,tex_name='x_{sh}')`
Is the `rsh+jsh`  the internal VSC losses ?  In Federico Milano’s book （Power System Modelling and Scripting）

**Comment by cuihantao** [comment]
It is lossless for now. Rsh and Xsh are the AC coupling impedance.

**Comment by YaeSakura1** [comment]
In VSC , the dc voltage droop control is also not implemented.

`self.psh = Algeb(info='active power injection into VSC', tex_name="P_{sh}", unit='p.u.', v_str='p0 * (mode_s0 + mode_s1)', e_str='u * (mode_s0 + mode_s1) * (p0 - psh) + ' 'u * (mode_s2 + mode_s3) * (v1 - v2 - vdc0)', diag_eps=True, )
self.qsh = Algeb(info='reactive power injection into VSC', tex_name="Q_{sh}", v_str='q0 * (mode_s0 + mode_s2)', e_str='u * (mode_s0 + mode_s2) * (q0 - qsh) + ' 'u * (mode_s1 + mode_s3) * (v0 - v)', diag_eps=True, )`

Is the psh `v_str` =p0, qsh `v_str`=q0 ?
I want to know when VSC_1 under vQ control mode and VSC_2 under PQ control mode the 'psh' and 'qsh' `e_str` is.

**Comment by cuihantao** [comment]
`psh` and `qsh` can be controlled at `p0` and `q0` but depends on the mode. 

Did you write the code above? It is different from what ANDES has in `vsc.py`.

```python
        self.qsh = Algeb(info='reactive power injection into VSC',
                         tex_name="Q_{sh}",
                         v_str='q0 * (mode_s0 + mode_s2)',
                         e_str='u * (mode_s0 + mode_s2) * (q0 - qsh) + '
                               'u * (mode_s1 + mode_s3) * (v0 - v)',
                         diag_eps=True,
                         )
```

Take this as an example, since Control method: 0-PQ, 1-PV, 2-vQ or 3-vV, `mode_s2` correspond to `vQ control`.  Therefore, the equation corresponding to the `qsh` variable is that `q0 - qsh = 0`, namely, controlling the reactive power injection to the setpoint `q0`

Please take a look the piecewise equation and consider the mode as 0/1 variables to enable only one part of the equation.

**Reply by YaeSakura1** [reply]
so the `psh` under vQ control mode , equation is `v1-v2-vdc0` ,is that right?

**Comment by cuihantao** [comment]
Yes. That is correct.

**Comment by YaeSakura1** [comment]
How do I look the Jacobian matrix after pflow?

**Comment by cuihantao** [comment]
Hi @YaeSakura1 .

Sorry but I will not be able to answer your questions if they have already been answered in the manual.

See https://docs.andes.app/en/stable/modeling.html#dae-storage for the Jacobian storage.

**Comment by YaeSakura1** [comment]
Hi,Dr. Cui,
Is it possible to implement electromechanical transient simulation of LCC-MMC HVDC in andes?

**Reply by cuihantao** [reply]
It is feasible. Please read the "Modeling Cookbook" and the source code of other implemented models.

**Reply by YaeSakura1** [reply]
Thanks!
Which part in the Modeling Cookbook should I read, and which model to look at ?

**Reply by cuihantao** [reply]
You might need to read and comprehend most of the Modeling Cookbook. I would also recommend checking out Power System Analysis Toolbox because it already has models for LCC HVDC.

**Comment by YaeSakura1** [comment]
Why does VSCShunt not have `self.flags.update({'tds':True})` but can do time domain simulation？

**Reply by cuihantao** [reply]
Power flow models participate in time-domain simulation by default. That's why we shared the same Line model for power flow and time-domain.

Static power flow models that have been replaced by dynamic models need to be turned off manually. See the `v_numeric(self, **kwargs)` function in `genbase.py` and `fload.py`. By the way, I converted several model source files into packages. You can browse these files in my `master` branch if you haven't updated.

**Reply by YaeSakura1** [reply]
Thank you very much!
How do I update your `master` branch to my branch?

**Comment by cuihantao** [comment]
There are multiple ways to do it. You can create a pull request to merge my master into your fork. See this link and related ones https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/about-pull-requests

Regards,
Hantao Cui

**Reply by YaeSakura1** [reply]
thank you!
also, how do I update my local files

**Comment by cuihantao** [comment]
You can do a `git pull` after merging the PR, assume you are on the same branch.

You might want to take a tutorial of Git and GitHub as these are general questions.

Regards,
Hantao Cui

**Reply by YaeSakura1** [reply]
I have created a PR

**Reply by YaeSakura1** [reply]
it said can’t automatically merge, is that mean i need to use the command line to merge ?

**Comment by YaeSakura1** [comment]
why i run the `andes prepare` command in pycharm , the following error has occurred:

- andes version：1.5.5.post16.dev0+ge946af25
- python version：3.9
- OS：windows 10
- Error Message:
 
> Traceback (most recent call last):
>   File "<string>", line 1, in   
> File "C:\Users\Administrator\anaconda3\envs\andes\lib\site-packages\multiprocess\spawn.py", line 116, in spawn_main
>     exitcode = _main(fd, parent_sentinel)
>   File "C:\Users\Administrator\anaconda3\envs\andes\lib\site-packages\multiprocess\spawn.py", line 125, in _main
>     prepare(preparation_data)
>   File "C:\Users\Administrator\anaconda3\envs\andes\lib\site-packages\multiprocess\spawn.py", line 236, in prepare
>     _fixup_main_from_path(data['init_main_from_path'])
>   File "C:\Users\Administrator\anaconda3\envs\andes\lib\site-packages\multiprocess\spawn.py", line 287, in _fixup_main_from_path
>     main_content = runpy.run_path(main_path,
> File "C:\Users\Administrator\anaconda3\envs\andes\lib\runpy.py", line 268, in run_path
>     return _run_module_code(code, init_globals, run_name,
>   File "C:\Users\Administrator\anaconda3\envs\andes\lib\runpy.py", line 97, in _run_module_code
>     _run_code(code, mod_globals, init_globals,
>   File "C:\Users\Administrator\anaconda3\envs\andes\lib\runpy.py", line 87, in _run_code
>     exec(code, run_globals)
>   File "C:\Users\Administrator\ANDES\andes\test.py", line 2, in <module>
>     andes.prepare()
>   File "C:\Users\Administrator\ANDES\andes\andes\main.py", line 740, in prepare
>     system.prepare(quick=quick, incremental=incremental, models=models,
>   File "C:\Users\Administrator\ANDES\andes\andes\system.py", line 345, in prepare
>     self._mp_prepare(models, quick, pycode_path, ncpu=ncpu)
> File "C:\Users\Administrator\ANDES\andes\andes\system.py", line 392, in _mp_prepare
>     Pool(ncpu).map(_prep_model, model_list)
>   File "C:\Users\Administrator\ANDES\andes\andes\utils\lazyimport.py", line 80, in __call__
>     return eval(self.__imported_name__)(*args, **kwargs)
>   File "C:\Users\Administrator\anaconda3\envs\andes\lib\site-packages\multiprocess\pool.py", line 212, in __init__
>     self._repopulate_pool()
>   File "C:\Users\Administrator\anaconda3\envs\andes\lib\site-packages\multiprocess\pool.py", line 303, in _repopulate_pool
>     return self._repopulate_pool_static(self._ctx, self.Process,
>   File "C:\Users\Administrator\anaconda3\envs\andes\lib\site-packages\multiprocess\pool.py", line 326, in _repopulate_pool_static
>     w.start()
>   File "C:\Users\Administrator\anaconda3\envs\andes\lib\site-packages\multiprocess\process.py", line 121, in start
>     self._popen = self._Popen(self)
>   File "C:\Users\Administrator\anaconda3\envs\andes\lib\site-packages\multiprocess\context.py", line 327, in _Popen
>     return Popen(process_obj)
>   File "C:\Users\Administrator\anaconda3\envs\andes\lib\site-packages\multiprocess\popen_spawn_win32.py", line 45, in __init__
>     prep_data = spawn.get_preparation_data(process_obj._name)
>   File "C:\Users\Administrator\anaconda3\envs\andes\lib\site-packages\multiprocess\spawn.py", line 154, in get_preparation_data
>     _check_not_importing_main()
>   File "C:\Users\Administrator\anaconda3\envs\andes\lib\site-packages\multiprocess\spawn.py", line 134, in _check_not_importing_main
>     raise RuntimeError('''
> RuntimeError: 
> 
>         An attempt has been made to start a new process before the
>         current process has finished its bootstrapping phase.
> 
>         This probably means that you are not using fork to start your
>         child processes and you have forgotten to use the proper idiom
>         in the main module:
> 
>             if __name__ == '__main__':
>                 freeze_support()
>                 ...
> 
>         The "freeze_support()" line can be omitted if the program
>         is not going to be frozen to produce an executable.

please help！

**Reply by cuihantao** [reply]
Sorry but I have no idea.

**Reply by YaeSakura1** [reply]
But I didn't seem to get this error when I run `andes prepare` from the command line

**Comment by YaeSakura1** [comment]
What does the last few lines of vscshunt.py mean？
`self.a.e_str = '-psh'`
`self.v.e_str = '-qsh'`
`self.v1.e_str = '-pdc / (v1 - v2)'`
`self.v2.e_str = 'pdc / (v1 - v2)'`

**Reply by cuihantao** [reply]
Please search for `e_str` in the modeling cookbook.

**Reply by YaeSakura1** [reply]
Yes, i konw the `e_str` is the the string representation of the equation , i want to konw what is it different with the `a` ,`V`, `V1`, `v2` in acdcbase.py?

**Comment by YaeSakura1** [comment]
Will master slave control and dc voltage droop control be added in the future？

**Reply by cuihantao** [reply]
It will eventually be added, but I don't have a timeline.
